redact.go

```diff
@@ -1,186 +1,323 @@
 package redact
 
 import (
-	"fmt"
-	"regexp"
-	"strconv"
 	"strings"
 )
 
 func All(input string) string {
-	input = Email(input)
-	input = DateOfBirth(input)
-	input = Phone(input)
-	input = SSN(input)
-	input = CreditCard(input)
-	return input
+	var matches []match
+	matches = append(matches, matchCreditCard(input)...)
+	matches = append(matches, matchEmail(input)...)
+	matches = append(matches, matchPhoneNum(input)...)
+	matches = append(matches, matchSSN(input)...)
+	matches = append(matches, matchDOB(input)...)
+	return redactMatches(input, matches)
 }
 
-func DateOfBirth(input string) string {
-	found := findDOB(input)
-	for _, item := range found {
-		input = strings.ReplaceAll(input, item, "[DOB REDACTED]")
+func redactMatches(input string, matches []match) string {
+	for _, match := range matches {
+		replace := input[match.InputIndex : match.InputIndex+match.Length]
+		input = strings.ReplaceAll(input, replace, strings.Repeat("*", match.Length))
 	}
 	return input
 }
-func findDOB(input string) (dates []string) {
-	dates = append(dates, datesWithDashes.FindAllString(input, 1000000)...)
-	dates = append(dates, datesWithSlashes.FindAllString(input, 1000000)...)
-	dates = append(dates, datesInUSFormat.FindAllString(input, 1000000)...)
-	dates = append(dates, datesInEUFormat.FindAllString(input, 1000000)...)
-
-	return dates
-}
 
-func Email(input string) string {
-	found := findEmails(input)
-	for _, item := range found {
-		// TODO: Display domain name
-		input = strings.ReplaceAll(input, item, "[EMAIL REDACTED]")
-	}
-	return input
-}
-func findEmails(input string) (emails []string) {
-	var temp string
-	var email bool
-	for i, character := range input {
-		if breakNotFound(character) {
-			// TODO: Avoid allocations
-			temp = fmt.Sprintf("%s%c", temp, character)
+func matchCreditCard(input string) (matches []match) {
+	var start int
+	var length int
+	var isCandidate bool
+	var total int
+	for i := 0; i < len(input)-1; i++ {
+		character := input[i]
+		if !isNumeric(character) {
+			if isCreditCard(length, input[start:start+length]) {
+				matches = appendMatches(matches, start, length)
+				length = 0
+				start = i + 1
+				total = 0
+				continue
+			}
+			if breakNotFound(character) && !isNumeric(input[i+1]) {
+				start = i + 1
+				length = 0
+				isCandidate = false
+				continue
+			}
+			length++
 		} else {
-			if email {
-				emails = append(emails, buildEmail(input, temp, i))
-				email = false
+			number := int(character - '0')
+			if isCandidate {
+				length++
+				total += number
+			} else {
+				isCandidate = true
+				start = i
+				total = number
 			}
-			temp = ""
+		}
+	}
+	if isNumeric(input[len(input)-1]) {
+		//length++ // TODO: test coverage
```

Comment:
There are several conditions (like this one) that are not currently necessary to pass the current test suite.